<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Estimate the Cause of Recurrent Vivax Malaria using Genetic Data • Pv3Rs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<script src="lightswitch.js"></script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Estimate the Cause of Recurrent Vivax Malaria using Genetic Data">
<meta name="description" content="Plot malaria parasite genetic data on two or more episodes. Compute per-person posterior probabilities that each Plasmodium vivax (Pv) recurrence is a recrudescence, relapse, or reinfection (3Rs) using per-person P. vivax genetic data on two or more episodes and a statistical model described in Taylor, Foo and White (2022, \doi{doi:10.1101/2022.11.23.22282669}). Plot per-recurrence posterior probabilities.">
<meta property="og:description" content="Plot malaria parasite genetic data on two or more episodes. Compute per-person posterior probabilities that each Plasmodium vivax (Pv) recurrence is a recrudescence, relapse, or reinfection (3Rs) using per-person P. vivax genetic data on two or more episodes and a statistical model described in Taylor, Foo and White (2022, \doi{doi:10.1101/2022.11.23.22282669}). Plot per-recurrence posterior probabilities.">
<meta property="og:image" content="https://aimeertaylor.github.io/Pv3Rs/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">Pv3Rs</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/aimeertaylor/Pv3Rs/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="pv3rs-">Pv3Rs <a class="anchor" aria-label="anchor" href="#pv3rs-"></a>
</h1>
</div>
<!-- badges: start -->

<p>An R package for <em>Plasmodium vivax</em> molecular correction via statistical genetic inference of</p>
<ul>
<li>Relapse</li>
<li>Recrudescence</li>
<li>Reinfection</li>
</ul>
<p>The core function, <code><a href="reference/compute_posterior.html">compute_posterior()</a></code>, computes per-person posterior probabilities of relapse, recrudescence, and reinfection (recurrence states) using <em>P. vivax</em> genetic data on two or more episodes. To fully understand the core function, in addition to reading this README in its entirety and the pre-print cited below, we recommend reading the <code><a href="articles/demonstrate-usage.html">vignette("demonstrate-usage", "Pv3Rs")</a></code> and <a href="https://aimeertaylor.github.io/Pv3Rs/articles/posterior-probabilities.html">Understand posterior probabilities</a>.</p>
<p>Two other important features:</p>
<ul>
<li><p><code><a href="reference/plot_data.html">plot_data()</a></code> visualises genetic data for molecular correction, regardless of the analytical method (e.g., <em>Plasmodium falciparum</em> data intended for analysis using a WHO match-counting algorithm).</p></li>
<li><p><code><a href="reference/plot_simplex.html">plot_simplex()</a></code> can be used to visualise per-recurrence probabilities of relapse, recrudescence, and reinfection, or any other probability triplet summing to one.</p></li>
</ul>
<div class="section level2">
<h2 id="please-be-aware-of-the-following-points">Please be aware of the following points!<a class="anchor" aria-label="anchor" href="#please-be-aware-of-the-following-points"></a>
</h2>
<p>The Pv3Rs R package is not yet peer-reviewed and thus liable to modification. The model is described in the preprint <a href="https://doi.org/10.1101/2022.11.23.22282669" class="external-link">Taylor, Foo &amp; White, 2022</a>, building on a prototype in <a href="https://doi.org/10.1038/s41467-019-13412-x" class="external-link">Taylor &amp; Watson et al. 2019</a>.</p>
<div class="section level3">
<h3 id="prior-considerations">Prior considerations:<a class="anchor" aria-label="anchor" href="#prior-considerations"></a>
</h3>
<ul>
<li><p>Genetic data are modelled using a Bayesian model, whose prior is ideally informative (in [2] priors were generated by a time-to-event model built by James Watson) because the cause of recurrent <em>P. vivax</em> malaria is not always identifiable from genetic data alone: when the data are consistent with recurrent parasites that are relatively unrelated to those in all preceding infections, both reinfection and relapse are plausible; meanwhile, when the data are compatible with recurrent parasites that are clones of those in the preceding infection, both recrudescence and relapse are plausible.</p></li>
<li><p>The main Pv3Rs function, <code><a href="reference/compute_posterior.html">compute_posterior()</a></code>, could be applied to <em>P. falciparum</em> by setting the prior probability of relapse to zero, but genotyping errors, which are not accounted for under the current Pv3Rs model, are liable to lead to the misclassification of recrudescence as reinfection when the prior probability of relapse is zero (and of recrudescence as relapse when the prior probability of relapse exceeds zero).</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="notable-assumptions-and-limitations">Notable assumptions and limitations:<a class="anchor" aria-label="anchor" href="#notable-assumptions-and-limitations"></a>
</h3>
<p>As with any model, Pv3Rs makes various assumptions that limit its capabilities in some settings.</p>
<div class="section level4">
<h4 id="mutually-exclusive-recurrent-states">Mutually exclusive recurrent states<a class="anchor" aria-label="anchor" href="#mutually-exclusive-recurrent-states"></a>
</h4>
<p>Recurrence states are modelled as mutually exclusive, suitable for studies where participants are actively followed up frequently and where all detected infections are treated to the extent that parasitaemia drops below some detectable level before recurrence, if recurrence occurs. In studies with untreated or accumulated infections, outputs may not be meaningful.</p>
</div>
<div class="section level4">
<h4 id="unmodelled-complexities">Unmodelled complexities<a class="anchor" aria-label="anchor" href="#unmodelled-complexities"></a>
</h4>
<p>We do not model all the complexities around molecular correction. For example, population structure, including household effects; failure to capture low-density clones in a blood sample of limited volume <a href="https://doi.org/10.1016/S0169-4758(98)01340-4" class="external-link">[Snounou &amp; Beck, 1998]</a>; and hidden biomass the spleen and bone marrow <a href="https://doi.org/10.1016/j.pt.2019.08.009" class="external-link">[Markus, 2019]</a>. Users must interpret outputs in context of the study and its methods. For example, we expect Pv3Rs to output probable relapse if a person is reinfected by a new mosquito but with parasites that are recently related to those that caused a previous infection, as might happen in household transmission chains.</p>
</div>
<div class="section level4">
<h4 id="sibling-misspecification">Sibling misspecification<a class="anchor" aria-label="anchor" href="#sibling-misspecification"></a>
</h4>
<p>Relapsing parasites that are siblings of parasites in previous infections can be meiotic, parent-child-like, regular or half siblings, but we model all sibling parasites as regular siblings via the following assumptions:</p>
<ul>
<li>Allele inheritance is independent (not true of meiotic siblings)</li>
<li>Aibling relationships are transitive (not true of parent-child-like trios or some half-sibling trios)</li>
<li>Alleles of a sibling cluster are drawn from at most two parental alleles (not true of half siblings)</li>
</ul>
<p>In our experience, half sibling misspecification leads to some misclassification of relapses as reinfections; see <a href="https://aimeertaylor.github.io/Pv3Rs/articles/half-siblings.pdf">Understand half-sibling misspecification</a>. A descriptive study to explore the extent of half-sibling misspecification is recommended (an example will be provided in an upcoming manuscript).</p>
</div>
<div class="section level4">
<h4 id="observation-errors-and-de-novo-mutations">Observation errors and <em>de novo</em> mutations<a class="anchor" aria-label="anchor" href="#observation-errors-and-de-novo-mutations"></a>
</h4>
<p>We do not model undetected alleles, genotyping errors, or <em>de novo</em> mutations. Recrudescent parasites are modelled as perfect clones under Pv3Rs. As such, the posterior probability of recrudescence is rendered zero by errors and mutations. This becomes more likely when there are data on more markers. Sensitivity analyses that explore the impact of errors and mutations on recurrence state probabilities are merited.</p>
</div>
<div class="section level4">
<h4 id="interpreting-probable-reinfection-and-recrudescence">Interpreting probable reinfection and recrudescence<a class="anchor" aria-label="anchor" href="#interpreting-probable-reinfection-and-recrudescence"></a>
</h4>
<p>When data are not sufficiently informative to distinguish between recrudescence and relapse (or reinfection and relapse), the posterior probabilities of recrudescence and relapse (or reinfection and relapse) are heavily influenced by a model assumption over relationship graphs; see <a href="https://aimeertaylor.github.io/Pv3Rs/articles/enumerate.pdf">Understand graph-prior ramifications</a>. The development of a more biologically-principled generative model on parasite relationships is merited.</p>
<table class="table">
<colgroup>
<col width="64%">
<col width="35%">
</colgroup>
<thead><tr class="header">
<th>Limitation</th>
<th>Reason</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Possible misclassification of persistent and/or accumulated states</td>
<td>Modelling recurrent states as mutually exclusive</td>
</tr>
<tr class="even">
<td>Possible inconsistency with data on more-and-more markers</td>
<td>Not modelling errors</td>
</tr>
<tr class="odd">
<td>Possible misclassification of relapse</td>
<td>Half-sibling misspecification and not modelling errors</td>
</tr>
<tr class="even">
<td>Possible misclassification of recrudescence</td>
<td>Not modelling errors</td>
</tr>
<tr class="odd">
<td>Possible misclassification of reinfection</td>
<td>Not modelling population structure</td>
</tr>
<tr class="even">
<td>Strong prior impact on posterior</td>
<td>Recurrent states are not always identifiable from genetic data alone</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="computational-limits">Computational limits:<a class="anchor" aria-label="anchor" href="#computational-limits"></a>
</h3>
<ul>
<li><p>Pv3Rs scales to hundreds of markers but not whole-genome sequence (WGS) data.</p></li>
<li><p>We do not recommend running <code><a href="reference/compute_posterior.html">compute_posterior()</a></code> for data whose total genotype count (sum of per-episode multiplicities of infection) exceeds eight. If the total genotype counts exceeds eight but there are multiple recurrences, it might be possible to compute posterior probabilities by analysing episodes pairwise (this approach was used in [2] and we’re working currently on an improved version).</p></li>
<li><p>The per-marker allele limit of <code><a href="reference/compute_posterior.html">compute_posterior()</a></code> is untested. Very high marker cardinalities could lead to very small allele frequencies and thus some underflow problems.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="population-level-allele-frequencies">Population-level allele frequencies:<a class="anchor" aria-label="anchor" href="#population-level-allele-frequencies"></a>
</h3>
<p>In addition to <em>P. vivax</em> allelic data on two or more episodes, <code><a href="reference/compute_posterior.html">compute_posterior()</a></code> requires as input population-level allele frequencies. To minimise bias due to within-host selection of recrudescent parasites, we recommend using only enrolment episodes to estimate population-level allele frequencies, and ideally enrolment episodes from study participants selected at random, not only study participants who experience recurrence. That said, if most recurrences are either reinfections or relapses, both of which are draws from the mosquito population (albeit a delayed draw in the case of a relapse), assuming there is no systematic within-patient selection (as might occur when infections encounter lingering drug pressure), estimates based on all episodes should be unbiased and more precise than those based on enrolment episodes only.</p>
</div>
<div class="section level3">
<h3 id="read-count-data">Read-count data:<a class="anchor" aria-label="anchor" href="#read-count-data"></a>
</h3>
<p>Unfortunately, the Pv3Rs model does not exploit data on read counts at present. However, read-count data could be used to compute population-level allele frequencies, assuming they are not biased by experimental artefacts.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install or update devtools from CRAN</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install Pv3Rs from GitHub </span></span>
<span><span class="co"># We recommend doing this in RStudio: RStudio installs pandoc, required for</span></span>
<span><span class="co"># vignette building. If not, you might need to install pandoc and check its</span></span>
<span><span class="co"># path; otherwise set build_vignettes = FALSE</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"aimeertaylor/Pv3Rs"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and attach Pv3Rs</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://aimeertaylor.github.io/Pv3Rs/">Pv3Rs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List links to all available documentation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"Pv3Rs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># List links to vignettes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"Pv3Rs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View function documentation, e.g., </span></span>
<span><span class="op">?</span><span class="va">compute_posterior</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/aimeertaylor/Pv3Rs/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/aimeertaylor/Pv3Rs/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Pv3Rs</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Aimee Taylor <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2337-8992" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Yong See Foo <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3010-9106" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>European Union, Project 101110393 <br><small class="roles"> Funder </small>   </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/aimeertaylor/Pv3Rs/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/aimeertaylor/Pv3Rs/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aimee Taylor, Yong See Foo, European Union, Project 101110393.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
