---
title: "Understand treatment of intra-episode siblings"
output: rmarkdown::html_vignette
description: "Expands upon a digression in **Understand posterior probabilities**"
vignette: >
  %\VignetteIndexEntry{Explore treatment of intra-episode siblings}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  out.width = "100%", # Line width taken up by plot
  fig.dpi = 300, # For quality 
  fig.width = 7, # Increase to zoom out
  fig.height = 4, 
  cache = FALSE
)
```

```{r setup, echo = FALSE}
library(Pv3Rs)
# Access non-exported function
recombine_parent_ids <- utils::getFromNamespace("recombine_parent_ids", "Pv3Rs") 
load("./vignette_data/maxima.rda")
```

This vignette expands a digression in 
[Understand posterior probabilities](https://aimeertaylor.github.io/Pv3Rs/articles/understand-posterior.html). 

`compute_posterior()` sums over graphs with cliques of three or more
intra-episode siblings whenever the `MOIs` used contain values that exceed
two. However, these graphs only acquire non-zero likelihoods if the user
specifies `MOIs` that also exceed the diversity in the data `y`.

In any case, when MOI estimates are based on bulk parasite genetic data (i.e.,
not single-cell data), groups of three or more non-half intra-episode siblings
collapse to groups of two (i.e., groups of three or more non-half intra-episode
siblings have no practical support) because:

- MOI estimates based on maximum per-marker allele counts cannot exceed two for
groups of three or more non-half siblings, because groups of non-half siblings 
can draw from at most two parental alleles per marker.

- MOI estimates based on heteroallelic marker counts likely don't exceed two because the
heteroallelic marker counts of groups of three or more non-half siblings cannot
exceed that of the two parental genotypes from which they derive.

# Example of a group of four intra-episode siblings collapsing to two

Before demonstrating how intra-episode sibling groups collapse to pairs, we
write a function to simulate data on 200 markers all with the same
user-specified allelic richness (marker cardinality) for an enrolment episode
comprising a stranger plus a group of four siblings, two from one oocyst, two
from another, all drawing from the same two unrelated parental genotypes, and a
recurrence with one sibling. Technically, the enrolment episode contains five
genetically distinct genotypes and thus has a MOI of five.

```{r}
simulate_data <- function(marker_cardinality){
  
  # Magic numbers / quantities
  set.seed(5) # For reproducibility
  n_markers <- 200 # Number of markers
  n_strangers <- 3 # Number of stranger parasites
  n_oocysts <- 2 # Number of oocysts to draw from
  
  # Derived quantities
  alleles <- letters[1:marker_cardinality]
  markers <- paste0("m", 1:n_markers) # Marker names
  
  # Uniform allele frequencies 
  fs <- sapply(markers, simplify = FALSE, 
               function(m) setNames(rep(1/marker_cardinality, marker_cardinality), alleles))
  
  # Sample strangers
  strangers <- sapply(1:n_strangers, function(i) {
    sapply(markers, function(t) sample(names(fs[[t]]), size = 1, prob = fs[[t]]))
  })
  
  # Designate strangers
  parents <- strangers[, 1:2]
  
  # Map the markers to chromosomes. Assume equally sized chromosomes â€” reasonable
  # if and only if we later assume an equal number of crossovers per chromosome
  chrs_per_marker <- round(seq(0.51, 14.5, length.out = n_markers))
  
  # Sample parental allocations dependently per-oocyst 
  cs <- lapply(1:n_oocysts, function(o) recombine_parent_ids(chrs_per_marker))
  
  # Construct children from parental allocations 
  all_children <- lapply(1:n_oocysts, function(o) {
    oocyst_chidren <- sapply(1:n_markers, function(i) {
      sapply(1:ncol(cs[[o]]), function(j) parents[i,cs[[o]][i,j]])
    })
    colnames(oocyst_chidren) <- markers
    return(oocyst_chidren)
  })
  
  # Make enrolment infection
  enrol <- apply(rbind(all_children[[1]][1:2,], 
                       all_children[[2]][1:2,], 
                       strangers[,3]), 2, unique, simplify = F)
  
  # Make paired data
  data <- list(enrol = enrol, recur = as.list(all_children[[1]][1,]))
  
  return(list(data = data, fs = fs))
}
```

However, MOI estimates based on maximum per-marker allele counts are 
three and one when markers are polyallelic:

```{r}
polyallelic <- simulate_data(10) 
determine_MOIs(polyallelic$data)
```

And two and one when markers are biallelic: 

```{r}
biallelic <- simulate_data(2) 
determine_MOIs(biallelic$data)
```

Suppose we estimate MOIs of 3 and 1 for the biallelic data using an external
software that exploits heteroallelic marker counts, and input those external
estimates into `compute_posterior()`. Providing data are simulated on a large
number of markers (200 above), we recover almost exactly the same posterior
probabilities using both the

- polyallelic data without user-specified MOIs
- biallelic data with user-specified MOIs of three and one

```{r}
ppost <- suppressMessages(compute_posterior(y = polyallelic$data, 
                                            fs = polyallelic$fs, 
                                            return.logp = T))
bpost <- suppressMessages(compute_posterior(y = biallelic$data, 
                                            fs = biallelic$fs, 
                                            MOIs = c(3,1), return.logp = T))

ppost$marg
bpost$marg
```

However, the following graphs, which have cliques of three intra-episode
siblings, have zero likelihood given the polyallelic data:

```{r echo = FALSE, fig.height = 1.5} 
MOIs <- c(3,1) # MOIs based on data 
gs <- paste0("g", 1:sum(MOIs)) # Genotype names
ts_per_gs <- rep(1:length(MOIs), MOIs) # Episode indices per genotype
llikes <- sapply(ppost$RGs, function(RG) RG$logp) # Extract log likelihoods
par(mar = c(0, 0.5, 0, 0.5), mfrow = c(1,5)) # Plotting parameters
for(RG in ppost$RGs[which(is.infinite(llikes))]) {
plot_RG(RG_to_igraph(RG, MOIs),
        vertex_palette = "Greys",
        edge.curved = 0.08,
        vertex.size = 40,
        edge.width = 3,
        vertex.frame.width = 2,
        vertex.label.cex = 1,
        vertex.label.color = "black")
}
```

Whereas, all relationships have non-zero likelihood given the biallelic data:

```{r} 
llikes <- sapply(bpost$RGs, function(RG) RG$logp) # Extract log likelihoods
any(is.infinite(llikes)) # Are there any minus infinity log likelihoods?
```

# Contribution to upper bounds 

The summation over graphs with cliques or three or more siblings is possibly
redundant given these graphs have no practical support. Even when these graphs
have zero likelihood, they contribute to maximum probabilities of recrudescence
/ reinfection through the conditionally uniform prior on graphs. However, for
upper bounds on the probability of reinfection / recrudescence given a single
recurrence the contribution is very small, especially for reinfection:

```{r echo = FALSE}
all_MOIs_chr <- colnames(maxima)
all_MOIs <- sapply(all_MOIs_chr, function(x) as.numeric(strsplit(x, split = "")[[1]]))
maxMOIs <- sapply(all_MOIs, function(x) max(x))
nMOIs <- sapply(all_MOIs, length) # Number of episodes
condition <- c(maxMOIs > 2 & nMOIs == 2) # MOIs > 2 (possibly of diff) and single recurrence
```

```{r echo = FALSE}
diffs <- maxima["I_with", condition] - maxima["I_wout", condition]
plot(NULL, xlim = c(0.5, 1), ylim = c(0.5,1), main = "Posterior bounds on reinfection", 
     xlab = "without summation", 
     ylab = "with summation"); abline(a = 0, b = 1)
points(y = maxima["I_with", condition], x = maxima["I_wout", condition], 
       pch = 21, bg = "red")
title(sub = sprintf("Largest difference: %s", round(max(abs(diffs)), 4))) 
```


```{r echo = FALSE}
diffs <- maxima["C_with", condition] - maxima["C_wout", condition]
max_diff <- max(abs(diffs))
cutoff <- 0.0175
graphs_with_big_diff <- names(all_MOIs)[condition][abs(diffs) > cutoff]
```

```{r echo = FALSE}
# Compare with and without summation
plot(NULL, xlim = c(0.5, 1), ylim = c(0.5,1), main = "Posterior bounds on recrudescence", 
     xlab = "without summation", ylab = "with summation"); abline(a = 0, b = 1)
points(y = maxima["C_with", condition][abs(diffs) > cutoff], 
       x = maxima["C_wout", condition][abs(diffs) > cutoff], pch = 1, cex = 2)
points(y = maxima["C_with", condition], x = maxima["C_wout", condition], 
       pch = 21, bg = "yellow")

title(sub = sprintf("Largest difference: %s", round(max_diff, 4))) 
```

For recrudescence, `r sum(abs(diffs) > cutoff)` MOI vectors (circled in the plot
above) result in an absolute difference greater than `r cutoff`. They are those
that feature large MOIs (MOI vectors 
`r sapply(graphs_with_big_diff, gsub,pattern = "", replacement = " ")`) and thus
those whose graph spaces include graphs with the largest cliques of
intra-episode siblings.

The contribution is likely even smaller for all vectors of three or more MOIs
that are computationally feasible under Pv3Rs, because their constituent MOIs
are necessarily less than 7 (recall that computationally feasible MOI vectors
are those that sum to at most eight).
