<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Demonstrate Pv3Rs usage • Pv3Rs</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Demonstrate Pv3Rs usage">
<meta name="description" content="Demonstrates Pv3Rs workflow for a single study participant">
<meta property="og:description" content="Demonstrates Pv3Rs workflow for a single study participant">
<meta property="og:image" content="https://aimeertaylor.github.io/Pv3Rs/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Pv3Rs</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/aimeertaylor/Pv3Rs/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Demonstrate Pv3Rs usage</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/aimeertaylor/Pv3Rs/blob/main/vignettes/demonstrate-usage.Rmd" class="external-link"><code>vignettes/demonstrate-usage.Rmd</code></a></small>
      <div class="d-none name"><code>demonstrate-usage.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates the basic <code>Pv3Rs</code> workflow for
a single study participant:</p>
<ol style="list-style-type: decimal">
<li>Plot data</li>
<li>Compute recurrence state posterior probabilities</li>
<li>Plot per-recurrence probabilities</li>
</ol>
<p>We also show how to explore relationship graphs and their log
likelihoods. For a general understanding of posterior probabilities, see
<a href="https://aimeertaylor.github.io/Pv3Rs/articles/posterior-probabilities.html">Understand
posterior probabilities</a>. A tutorial demonstrating the
<code>Pv3Rs</code> workflow at the study-level is planned; we plan to
cover</p>
<ul>
<li>Generating pairwise probabilities when the total genotype count over
multiple recurrences exceeds eight</li>
<li>Computing false discovery rates</li>
<li>Sensitivity analysis for genotyping errors</li>
<li>Sensitivity analysis for sibling misspecification</li>
<li>Alternative approach using genetic distance and whole-genome
sequence data</li>
<li>Using posterior probabilities (rather than categorical
classifications) in downstream analyses</li>
<li>Estimating false negative rates</li>
</ul>
<div class="section level2">
<h2 id="basic-workflow">Basic workflow<a class="anchor" aria-label="anchor" href="#basic-workflow"></a>
</h2>
<p>We begin with a synthetic example of three episodes (episode names
are optional) with three markers (marker names are obligatory) whose
alleles have known frequencies, <code>fs</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Enrolment"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'b'</span>,<span class="st">'c'</span>,<span class="st">'d'</span><span class="op">)</span>, </span>
<span>                              m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'a'</span>,<span class="st">'b'</span><span class="op">)</span>,</span>
<span>                              m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'b'</span>,<span class="st">'c'</span>,<span class="st">'d'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          <span class="st">"Recurrence 1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'b'</span>,<span class="st">'d'</span><span class="op">)</span>, </span>
<span>                                m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'a'</span><span class="op">)</span>, </span>
<span>                                m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'a'</span>,<span class="st">'b'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          <span class="st">"Recurrence 2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'d'</span><span class="op">)</span>, </span>
<span>                                m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'a'</span><span class="op">)</span>, </span>
<span>                                m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'a'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0.27</span>, b <span class="op">=</span> <span class="fl">0.35</span>, c <span class="op">=</span> <span class="fl">0.18</span>, d <span class="op">=</span> <span class="fl">0.20</span><span class="op">)</span>,</span>
<span>           m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0.78</span>, b <span class="op">=</span> <span class="fl">0.14</span>, c <span class="op">=</span> <span class="fl">0.07</span>, d <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,</span>
<span>           m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0.21</span>, b <span class="op">=</span> <span class="fl">0.45</span>, c <span class="op">=</span> <span class="fl">0.26</span>, d <span class="op">=</span> <span class="fl">0.08</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="plot-data">1) Plot data<a class="anchor" aria-label="anchor" href="#plot-data"></a>
</h3>
<p>We plot the data using <code><a href="../reference/plot_data.html">plot_data()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_data.html">plot_data</a></span><span class="op">(</span>ys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Participant data"</span> <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, fs <span class="op">=</span> <span class="va">fs</span><span class="op">)</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>For each marker, different colours represent different alleles. The
legend below the main grid shows per-marker allele frequencies via
colour proportions, with one row per marker, ordered as in the main
grid; e.g., <code>d</code> at <code>m2</code> is rare.</p>
<p>The most parsimonious MOI estimates compatible with the above data,
3, 2, 1, are computed using <code><a href="../reference/determine_MOIs.html">determine_MOIs()</a></code>. The total
genotype count (sum over MOIs) is 6.</p>
<p><strong>Aside</strong> In this synthetic example, markers are
quart-allelic, imposing low upper bounds on MOI estimates based on
maximum per-marker allele counts. More diverse markers are recommended
for MOI estimation and recurrence state inference.</p>
</div>
<div class="section level3">
<h3 id="compute-recurrence-state-posterior-probabilities">2) Compute recurrence state posterior probabilities<a class="anchor" aria-label="anchor" href="#compute-recurrence-state-posterior-probabilities"></a>
</h3>
<p>The bulk of the computational time lies in computing log-likelihoods
of graphs of relationships between genotypes. The number of graphs
depends on the MOIs. By default, <code>compute_posterior</code> uses MOI
estimates generated by <code>determine_MOIs</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_posterior.html">compute_posterior</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">fs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of valid relationship graphs (RGs) is 1315</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Computing log p(Y|RG) for 1315 RGs</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Finding log-likelihood of each vector of recurrence states</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span></code></pre></div>
<p><strong>Aside</strong> We do not recommend running
<code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> for data whose total genotype count
(sum over MOIs) exceeds eight. That said, we have not encoded a hard
limit. In our experience, it is possible, but very long, to generate
posterior probabilities using data with a total genotype count up to 10;
above 10, calls to <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> are liable cause
memory-use problems and fail.</p>
<p>In the call to <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> above we did not
specify a prior, and so by default all three recurrence states were
assumed equally likely per recurrence.</p>
<p>Posterior probabilities of recurrent state sequences (where
<code>C</code> is recrudescence, <code>L</code> is relapse, and
<code>I</code> is reinfection) are stored in <code>post$joint</code>.
Here, we find the most likely sequence of recurrence states is IC with
posterior probability 0.4568:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">joint</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt;         IC         LC         IL         LL         II         LI         CC </span></span>
<span><span class="co">#&gt; 0.45682305 0.21338194 0.15385932 0.08501504 0.05986209 0.03105856 0.00000000 </span></span>
<span><span class="co">#&gt;         CL         CI </span></span>
<span><span class="co">#&gt; 0.00000000 0.00000000</span></span></code></pre></div>
<p>Per-recurrence posterior probabilities of recrudescence
<code>C</code>, relapse <code>L</code>, and reinfection <code>I</code>
are stored in <code>post$marg</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span><span class="op">$</span><span class="va">marg</span></span>
<span><span class="co">#&gt;                     C         L          I</span></span>
<span><span class="co">#&gt; Recurrence 1 0.000000 0.3294555 0.67054446</span></span>
<span><span class="co">#&gt; Recurrence 2 0.670205 0.2388744 0.09092065</span></span></code></pre></div>
<p>We refer to per-recurrence probabilities as <code>marg</code>
(shorthand for marginal) because they are computed by simply summing
over state sequence probabilities. For example, the probability of
<code>L</code> at <code>Recurrence 1</code> above is the sum of
probabilities over <code>LC</code>, <code>LL</code> and
<code>LI</code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span><span class="op">$</span><span class="va">joint</span><span class="op">[</span><span class="st">"LC"</span><span class="op">]</span> <span class="op">+</span> <span class="va">post</span><span class="op">$</span><span class="va">joint</span><span class="op">[</span><span class="st">"LL"</span><span class="op">]</span> <span class="op">+</span> <span class="va">post</span><span class="op">$</span><span class="va">joint</span><span class="op">[</span><span class="st">"LI"</span><span class="op">]</span></span>
<span><span class="co">#&gt;        LC </span></span>
<span><span class="co">#&gt; 0.3294555</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-per-recurrence-probabilities">3) Plot per-recurrence probabilities<a class="anchor" aria-label="anchor" href="#plot-per-recurrence-probabilities"></a>
</h3>
<p>Per-recurrence posterior probabilities are plotted on the simplex
using <code><a href="../reference/plot_simplex.html">plot_simplex()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oldpar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>no.readonly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_simplex.html">plot_simplex</a></span><span class="op">(</span>p.coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">marg</span>, Prior <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">oldpar</span><span class="op">)</span></span></code></pre></div>
<p>The point in the yellow region is most likely a recrudescence with
posterior probability greater than 0.5 (it falls in the bright yellow
region); the point in the red region is most likely a reinfection with
posterior probability greater than 0.5 (it falls in the bright red
region).</p>
</div>
</div>
<div class="section level2">
<h2 id="exploration-of-relationship-graphs">Exploration of relationship graphs<a class="anchor" aria-label="anchor" href="#exploration-of-relationship-graphs"></a>
</h2>
<p>To explore relationship graphs (RGs) and their log-likelihoods, set
<code>return.RG</code> and <code>return.logp</code> to
<code>TRUE</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_posterior.html">compute_posterior</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">fs</span>, return.RG <span class="op">=</span> <span class="cn">TRUE</span>, return.logp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of valid relationship graphs (RGs) is 1315</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Computing log p(Y|RG) for 1315 RGs</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Finding log-likelihood of each vector of recurrence states</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span></code></pre></div>
<p>We recover the same posterior as before.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">joint</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt;         IC         LC         IL         LL         II         LI         CC </span></span>
<span><span class="co">#&gt; 0.45682305 0.21338194 0.15385932 0.08501504 0.05986209 0.03105856 0.00000000 </span></span>
<span><span class="co">#&gt;         CL         CI </span></span>
<span><span class="co">#&gt; 0.00000000 0.00000000</span></span></code></pre></div>
<p>But the compute time was longer: to compute the posterior, summations
over per-marker allelic assignments that are equivalent up to
within-episode genotype permutations are redundant. As such, by default,
<code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> does not sum over them, conserving both
memory and compute time. The exploitation of permutation symmetry
requires a scheme to choose a single representative among permutations
that are otherwise equivalent. To compute meaningful graph likelihood
values (values that do not depend on the representative-choosing
scheme), all permutations are summed over when
<code>return.logp = TRUE</code>, increasing compute time, especially
when MOIs are large. Also, when user-specified MOIs exceed those of
<code><a href="../reference/determine_MOIs.html">determine_MOIs()</a></code>, all permutations are summed over because
the representative-choosing scheme is too complicated.</p>
<p>The log-likelihood of each relationship graph is returned. We plot
the relationship graph(s) with the largest likelihood. In this example,
there are two graphs with maximum likelihood; they are isomorphic up to
within-episode genotype permutations.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co"># Extract all log likelihoods</span></span>
<span><span class="va">llikes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">RGs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">RG</span><span class="op">)</span> <span class="va">RG</span><span class="op">$</span><span class="va">logp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get maximum log likelihood</span></span>
<span><span class="va">mllikes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">llikes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the relationship graphs (RGs) with the maximum log likelihood</span></span>
<span><span class="va">RGs</span> <span class="op">&lt;-</span> <span class="va">post</span><span class="op">$</span><span class="va">RGs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">llikes</span> <span class="op">-</span> <span class="va">mllikes</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot RGs with maximum log likelihoods</span></span>
<span><span class="va">oldpar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>no.readonly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Store user's options</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">4</span><span class="op">)</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">RGs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/plot_RG.html">plot_RG</a></span><span class="op">(</span><span class="fu"><a href="../reference/RG_to_igraph.html">RG_to_igraph</a></span><span class="op">(</span><span class="va">RGs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="../reference/determine_MOIs.html">determine_MOIs</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>, vertex.size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/box.html" class="external-link">box</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Add a legend</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottomright"</span>, pch <span class="op">=</span> <span class="fl">21</span>, </span>
<span>       pt.bg <span class="op">=</span>  <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">8</span>, <span class="st">"Set2"</span><span class="op">)</span> <span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">]</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, title <span class="op">=</span> <span class="st">"Episode"</span><span class="op">)</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">oldpar</span><span class="op">)</span> <span class="co"># Restore user's options</span></span></code></pre></div>
<p>Using log likelihoods, we can also find the equivalence class for
which the data are most probable when all relationship graphs in that
class are summed over.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># In the following code, we place two graphs in the same equivalence class if</span></span>
<span><span class="co"># they share the same likelihood. This is not ideal (two graphs that are not</span></span>
<span><span class="co"># isomorphic up to permutation could share the same likelihood), but it works</span></span>
<span><span class="co"># here: the plot shows only isomorphic graphs within the equivalence class.</span></span>
<span></span>
<span><span class="va">sorted_llikes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">llikes</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="co"># Sort log likelihoods</span></span>
<span><span class="va">adj_equal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">sorted_llikes</span>, lag <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span> <span class="co"># Find matches</span></span>
<span><span class="va">decr_idxs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">adj_equal</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># Change points: 2, 8, 14, 20, 32, ...</span></span>
<span><span class="va">class_sizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">decr_idxs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">decr_idxs</span><span class="op">)</span><span class="op">)</span> <span class="co"># Number of graphs per class</span></span>
<span></span>
<span><span class="co"># log likelihood of representative from each 'equivalence class' (EC)</span></span>
<span><span class="va">llikes_unique</span> <span class="op">&lt;-</span> <span class="va">sorted_llikes</span><span class="op">[</span><span class="va">decr_idxs</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># EC likelihood</span></span>
<span><span class="va">class_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">llikes_unique</span><span class="op">)</span><span class="op">*</span><span class="va">class_sizes</span></span>
<span><span class="va">max_class_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">class_ps</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">class_ps</span><span class="op">)</span><span class="op">)</span> <span class="co"># ML EC index </span></span>
<span><span class="va">max_idx</span> <span class="op">&lt;-</span> <span class="va">decr_idxs</span><span class="op">[</span><span class="va">max_class_p</span><span class="op">]</span> <span class="co"># Index of last graph in ML EC</span></span>
<span><span class="va">max_size</span> <span class="op">&lt;-</span> <span class="va">class_sizes</span><span class="op">[</span><span class="va">max_class_p</span><span class="op">]</span> <span class="co"># Number of graphs in ML EC</span></span>
<span></span>
<span><span class="co"># Plot all graphs within the ML EC </span></span>
<span><span class="va">oldpar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>no.readonly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># Store user's options</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">4</span><span class="op">)</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">RG_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">llikes</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="co"># order RGs by logl</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="op">(</span><span class="va">max_idx</span><span class="op">-</span><span class="va">max_size</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="va">max_idx</span><span class="op">)</span> <span class="op">{</span> <span class="co"># EC consists of the RGs with logl rank 21-32</span></span>
<span>  <span class="va">RG</span> <span class="op">&lt;-</span> <span class="va">post</span><span class="op">$</span><span class="va">RGs</span><span class="op">[[</span><span class="va">RG_order</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">RG_igraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RG_to_igraph.html">RG_to_igraph</a></span><span class="op">(</span><span class="va">RG</span>, <span class="fu"><a href="../reference/determine_MOIs.html">determine_MOIs</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/plot_RG.html">plot_RG</a></span><span class="op">(</span><span class="va">RG_igraph</span>, vertex.size <span class="op">=</span> <span class="fl">25</span>, vertex.label <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/box.html" class="external-link">box</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-12-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span><span class="va">oldpar</span><span class="op">)</span> <span class="co"># Restore user's options</span></span></code></pre></div>
<p><strong>Important considerations</strong></p>
<ul>
<li><p>The maximum-likelihood graph(s) might not be in the equivalence
class for which the data are most probable when all relationship graphs
in that class are summed over (true of the above example).</p></li>
<li><p>The maximum-likelihood graph(s) might be incompatible with the
maximum-posterior state sequence, (not true of the above example; recall
the most-likely sequence was IC).
<!-- see example in DevFiles/Investigate_Symmetries --></p></li>
<li><p>The graphs in the maximum-likelihood equivalence class might be
incompatible with the maximum-posterior state sequence (true of the
above example; the class with the largest likelihood contains graphs
with sibling edges that are incompatible with IC).</p></li>
</ul>
<!-- We suspect one of these cases above has to be false, but we cannot rule out
the possibility that both are true without further exploration -->
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aimee Taylor, Yong See Foo, European Union, Project 101110393.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
