<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Explore treatment of intra-episode siblings • Pv3Rs</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Explore treatment of intra-episode siblings">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Pv3Rs</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/demonstrate-usage.html">Demonstrate Pv3Rs usage</a></li>
    <li><a class="dropdown-item" href="../articles/enumerate.html">Enumerate graphs and partitions</a></li>
    <li><a class="dropdown-item" href="../articles/intra-episode-siblings.html">Explore treatment of intra-episode siblings</a></li>
    <li><a class="dropdown-item" href="../articles/understand-graph-prior.html">Understand graph prior ramifications</a></li>
    <li><a class="dropdown-item" href="../articles/understand-half-sibs.html">Understand half-sibling misspecification</a></li>
    <li><a class="dropdown-item" href="../articles/understand-posterior.html">Understand posterior estimates</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/aimeertaylor/Pv3Rs/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Explore treatment of intra-episode siblings</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/aimeertaylor/Pv3Rs/blob/main/vignettes/intra-episode-siblings.Rmd" class="external-link"><code>vignettes/intra-episode-siblings.Rmd</code></a></small>
      <div class="d-none name"><code>intra-episode-siblings.Rmd</code></div>
    </div>

    
    
<p>As the following example illustrates, Pv3Rs sums over graphs with
cliques of three or more intra-episode siblings when MOIs used by the
Pv3Rs model exceed two. However, when the data input via the
<code>y</code> argument of <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> are as
diverse as the MOIs used by the Pv3Rs model, graphs with cliques of
three or more intra-episode sibling parasites have zero likelihood; when
MOIs used by the Pv3Rs model are more diverse than the data input via
the <code>y</code> argument of <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> (e.g.,
because the user inputs elevated MOIs into
<code><a href="../reference/compute_posterior.html">compute_posterior()</a></code>), graphs with cliques of three or more
intra-episode siblings have non-zero likelihoods. In either case,
provided MOIs used by the Pv3Rs model are based on estimates from
per-episode parasite genetic data generated and analysed in bulk (i.e.,
not single-cell data), groups of three or more not-half intra-episode
siblings collapse to groups of two because</p>
<ul>
<li><p>MOI estimates based on maximum per-marker allele counts are at
most two for groups of three or more not-half siblings because groups of
not-half siblings can draw from at most two parental alleles per
marker.</p></li>
<li><p>MOI estimates based on heteroallelic marker counts count at most
two for groups of three or more not-half siblings because data from a
group of not-half siblings can be at most as diverse across markers as
the two parental genotypes from which they draw.</p></li>
</ul>
<div class="section level3">
<h3 id="example-of-a-group-of-four-intra-episode-siblings-collapsing-to-two">Example of a group of four intra-episode siblings collapsing to
two<a class="anchor" aria-label="anchor" href="#example-of-a-group-of-four-intra-episode-siblings-collapsing-to-two"></a>
</h3>
<p>Before showing how groups of intra-episode siblings collapse to pairs
of intra-episode siblings, we write a function to simulate data of a
given allelic richness (marker cardinality) on an enrolment episode
comprising a stranger plus a group of four siblings, two from one
oocyst, two from another, all drawing from the same two unrelated
parental genotypes, and a recurrence with one sibling. Technically, the
enrolment episode contains five genetically distinct genotypes and thus
has a MOI of five.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simulate_data</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">marker_cardinality</span><span class="op">)</span><span class="op">{</span></span>
<span>  </span>
<span>  <span class="co"># Magic numbers / quantities</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span> <span class="co"># For reproducibility</span></span>
<span>  <span class="va">n_markers</span> <span class="op">&lt;-</span> <span class="fl">200</span> <span class="co"># Number of markers</span></span>
<span>  <span class="va">n_strangers</span> <span class="op">&lt;-</span> <span class="fl">3</span> <span class="co"># Number of stranger parasites</span></span>
<span>  <span class="va">n_oocysts</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="co"># Number of oocysts to draw from</span></span>
<span>  </span>
<span>  <span class="co"># Derived quantities</span></span>
<span>  <span class="va">alleles</span> <span class="op">&lt;-</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">marker_cardinality</span><span class="op">]</span></span>
<span>  <span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"m"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_markers</span><span class="op">)</span> <span class="co"># Marker names</span></span>
<span>  </span>
<span>  <span class="co"># Uniform allele frequencies </span></span>
<span>  <span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">markers</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>               <span class="kw">function</span><span class="op">(</span><span class="va">m</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="va">marker_cardinality</span>, <span class="va">marker_cardinality</span><span class="op">)</span>, <span class="va">alleles</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Sample strangers</span></span>
<span>  <span class="va">strangers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_strangers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">t</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fs</span><span class="op">[[</span><span class="va">t</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1</span>, prob <span class="op">=</span> <span class="va">fs</span><span class="op">[[</span><span class="va">t</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Designate strangers</span></span>
<span>  <span class="va">parents</span> <span class="op">&lt;-</span> <span class="va">strangers</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="co"># Map the markers to chromosomes. Assume equally sized chromosomes — reasonable</span></span>
<span>  <span class="co"># if and only if we later assume an equal number of crossovers per chromosome</span></span>
<span>  <span class="va">chrs_per_marker</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0.51</span>, <span class="fl">14.5</span>, length.out <span class="op">=</span> <span class="va">n_markers</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Sample parental allocations dependently per-oocyst </span></span>
<span>  <span class="va">cs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_oocysts</span>, <span class="kw">function</span><span class="op">(</span><span class="va">o</span><span class="op">)</span> <span class="fu">recombine_parent_ids</span><span class="op">(</span><span class="va">chrs_per_marker</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Construct children from parental allocations </span></span>
<span>  <span class="va">all_children</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_oocysts</span>, <span class="kw">function</span><span class="op">(</span><span class="va">o</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">oocyst_chidren</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">n_markers</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cs</span><span class="op">[[</span><span class="va">o</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">j</span><span class="op">)</span> <span class="va">parents</span><span class="op">[</span><span class="va">i</span>,<span class="va">cs</span><span class="op">[[</span><span class="va">o</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">oocyst_chidren</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">markers</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">oocyst_chidren</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Make enrolment infection</span></span>
<span>  <span class="va">enrol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">all_children</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span>, </span>
<span>                       <span class="va">all_children</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>,<span class="op">]</span>, </span>
<span>                       <span class="va">strangers</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span>, <span class="fl">2</span>, <span class="va">unique</span>, simplify <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># Make paired data</span></span>
<span>  <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>enrol <span class="op">=</span> <span class="va">enrol</span>, recur <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">all_children</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span>, fs <span class="op">=</span> <span class="va">fs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>However, MOI estimates based on maximum per-marker allele counts are
three and one when markers are polyallelic:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">polyallelic</span> <span class="op">&lt;-</span> <span class="fu">simulate_data</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/determine_MOIs.html">determine_MOIs</a></span><span class="op">(</span><span class="va">polyallelic</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3 1</span></span></code></pre></div>
<p>And two and one when markers are biallelic:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biallelic</span> <span class="op">&lt;-</span> <span class="fu">simulate_data</span><span class="op">(</span><span class="fl">2</span><span class="op">)</span> </span>
<span><span class="fu"><a href="../reference/determine_MOIs.html">determine_MOIs</a></span><span class="op">(</span><span class="va">biallelic</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2 1</span></span></code></pre></div>
<p>Suppose we estimate MOIs of 3 and 1 for the biallelic data using an
external software that exploits heteroallelic marker counts (rather than
the maximum number of per-marker alleles), and input those external
estimates into <code>compute_posterior</code>. Providing data are
simulated on a large number of markers (200 above), we recover almost
exactly the same posterior probabilities using both the</p>
<ul>
<li>polyallelic data without user-specified MOIs</li>
<li>biallelic data with user-specified MOIs three and one</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ppost</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="../reference/compute_posterior.html">compute_posterior</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">polyallelic</span><span class="op">$</span><span class="va">data</span>, </span>
<span>                                            fs <span class="op">=</span> <span class="va">polyallelic</span><span class="op">$</span><span class="va">fs</span>, </span>
<span>                                            return.logp <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">bpost</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="../reference/compute_posterior.html">compute_posterior</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">biallelic</span><span class="op">$</span><span class="va">data</span>, </span>
<span>                                            fs <span class="op">=</span> <span class="va">biallelic</span><span class="op">$</span><span class="va">fs</span>, </span>
<span>                                            MOIs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span>, return.logp <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ppost</span><span class="op">$</span><span class="va">marg</span></span>
<span><span class="co">#&gt;               C         L             I</span></span>
<span><span class="co">#&gt; recur 0.6666667 0.3333333 2.708985e-116</span></span>
<span><span class="va">bpost</span><span class="op">$</span><span class="va">marg</span></span>
<span><span class="co">#&gt;               C         L            I</span></span>
<span><span class="co">#&gt; recur 0.6666663 0.3333337 4.625927e-18</span></span></code></pre></div>
<p>However, given the polyallelic data the likelihoods of the following
graphs with cliques of three intra-episode siblings are zero:</p>
<p><img src="intra-episode-siblings_files/figure-html/unnamed-chunk-6-1.png" width="100%"></p>
<p>Whereas, all relationships have non-zero likelihoods given the
biallelic data:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">llikes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">bpost</span><span class="op">$</span><span class="va">RGs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">RG</span><span class="op">)</span> <span class="va">RG</span><span class="op">$</span><span class="va">logp</span><span class="op">)</span> <span class="co"># Extract log likelihoods</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/is.finite.html" class="external-link">is.infinite</a></span><span class="op">(</span><span class="va">llikes</span><span class="op">)</span><span class="op">)</span> <span class="co"># Are there any minus infinity log likelihoods?</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="contribution-to-upper-bounds">Contribution to upper bounds<a class="anchor" aria-label="anchor" href="#contribution-to-upper-bounds"></a>
</h2>
<p>The summation over graphs with cliques or three or more siblings is
possibly redundant given these graphs have no practical support. Even
when these graphs have zero likelihood, they contribute to maximum
probabilities of recrudescence / reinfection through the conditionally
uniform prior on graphs. However, for upper bounds on the probability of
reinfection / recrudescence given a single recurrence the contribution
is very small, especially for reinfection:</p>
<p><img src="intra-episode-siblings_files/figure-html/unnamed-chunk-9-1.png" width="100%"></p>
<p>For recrudescence, 3 MOI vectors (circled below) result in an
absolute difference greater than 0.0175. They are those that feature
large MOIs (MOI vectors 52, 62, 71) and thus those whose graph spaces
include graphs with the largest cliques of intra-episode siblings.</p>
<p><img src="intra-episode-siblings_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
<p>The contribution is likely smaller for all vectors of three or more
MOIs that are computationally feasible under Pv3Rs because their
constituent MOIs are necessarily less than 7 (computationally feasible
MOI vectors are those where the total genotype count is at most
eight).</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aimee Taylor, Yong See Foo, European Union, Project 101110393.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
